compiler/rustc_mir_transform/src/const_prop_lint.rs

```diff
@@ -286,25 +286,27 @@ impl<'mir, 'tcx> ConstPropagator<'mir, 'tcx> {
     }
 
     /// Returns the value, if any, of evaluating `c`.
-    fn eval_constant(
-        &mut self,
-        c: &Constant<'tcx>,
-        _source_info: SourceInfo,
-    ) -> Option<OpTy<'tcx>> {
+    fn eval_constant(&mut self, c: &Constant<'tcx>, source_info: SourceInfo) -> Option<OpTy<'tcx>> {
         // FIXME we need to revisit this for #67176
         if c.needs_subst() {
             return None;
         }
 
-        match self.ecx.const_to_op(&c.literal, None) {
-            Ok(op) => Some(op),
-            Err(error) => {
-                let tcx = self.ecx.tcx.at(c.span);
-                let err = ConstEvalErr::new(&self.ecx, error, Some(c.span));
-                err.report_as_error(tcx, "erroneous constant used");
-                None
+        let res = self.use_ecx(source_info, |this| this.ecx.const_to_op(&c.literal, None));
+        if res.is_none() {
```

Comment:
we do need the error to avoid emitting the note in case of TooGeneric and similar "nonfatal const prop failures"

Reply:
Urgh but that's a terrible abstraction violation. :/

Reply:
it's what `report_as_error` was doing before. So you could add a `maybe_report_erroneous_const()` method to the error type and call it where we called `report_as_error` before